### PR TITLE
Feature/deequ 149 uniqueness

### PR DESCRIPTION
**Pull request by rambrus**
*Issue #, if available:*
https://github.com/awslabs/deequ/issues/149

*Description of changes:*
@sscdotopen This is a draft PR for adding the filter capability to `Uniqueness` analyzer.

I found that adding this capability to the descendants of [ScanShareableFrequencyBasedAnalyzer](https://github.com/awslabs/deequ/blob/master/src/main/scala/com/amazon/deequ/analyzers/GroupingAnalyzers.scala#L83) class is not so straightforward, since there is a performance optimization in [AnalysisRunner](https://github.com/awslabs/deequ/blob/master/src/main/scala/com/amazon/deequ/analyzers/runners/AnalysisRunner.scala#L175) for this type of analyzers. As I understand, the grouped state of the dataset is shared between analyzers that are grouping the data by the same columns.

I think this optimization cannot be applied when `filterCondition` is available for an Analyzer, since filtering is messing up the `Frequencies` and `NumRows`.
So, in my understanding, `GroupingAnalyzers` with filter conditions need to be processed separately.

What do you think? Do you have any hint for the implementation?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by sscdotopen** [comment]
You are absolutely right, filtering for grouping analyzers is harder. Our AnalysisRunner generates the queries for the state computation in a way that it tries to minimize the overall number of scans over the data. For GroupingAnalyzers, we try to group once and run the aggregation on the grouped results for all analyzers that use the same partitioning (e.g., group by the same set of attributes).

This is affected when we filter before the grouping (as you correctly point out). A potential problem of your PR is (if I understand it right), that you execute every filtering group by on its own, which can be come very slow. We would need to make sure that we execute all analyzers that filter and group by in the same way together.

Does that make sense?

**Comment by rambrus** [comment]
Yes indeed, it executes every filtering group by on its own. Just wanted to add a quick and dirty implementation to make sure the whole concept is working.
Makes sense, I'll take a look at the possibilities of executing the analyzers together that are filter and group the data the same way. 

Thanks for the hints, will get back to you with the updates!

**Comment by sscdotopen** [comment]
Looking forward to the update! And thank you for taking on such a hard problem!

**Comment by rambrus** [comment]
@sscdotopen I completed the updates, please take a look and let me know your thoughts.

*Summary of the changes:*
- Added the filter support to [FrequencyBasedAnalyzer] (https://github.com/awslabs/deequ/pull/198/files#diff-f94e52c979477fa69d85eb9caebe64a9)
- Altered the processing of grouping analyzers in [AnalysisRunner](https://github.com/awslabs/deequ/blob/7de97f946cd250c1d1751062733881e1d8b11af0/src/main/scala/com/amazon/deequ/analyzers/runners/AnalysisRunner.scala#L175), now the analyzers are grouped by `grouping columns` and the `filter condition`
- Added to filter support to [Uniqueness](https://github.com/awslabs/deequ/blob/7de97f946cd250c1d1751062733881e1d8b11af0/src/main/scala/com/amazon/deequ/analyzers/runners/AnalysisRunner.scala#L175) and [UniqueValueRatio](https://github.com/awslabs/deequ/blob/7de97f946cd250c1d1751062733881e1d8b11af0/src/main/scala/com/amazon/deequ/analyzers/UniqueValueRatio.scala) analyzers
- Updated the related Checks and test cases

**Comment by rambrus** [comment]
> This looks really great, I only have a few cosmetic change requests to your code.
> 
> I have one more favor to ask. Could add a test that controls the number of spark jobs that are generated? We should make sure that analyzers that use the same grouping and filtering are executed together (I think this is the case in your code, but I would prefer to have a test for it).
> 
> Here is an example of such a test: https://github.com/awslabs/deequ/blob/master/src/test/scala/com/amazon/deequ/analyzers/runners/AnalysisRunnerTests.scala#L50

Sure, makes sense, will do

File: src/main/scala/com/amazon/deequ/analyzers/GroupingAnalyzers.scala [hunk 1]
```diff
@@ -66,17 +67,23 @@ object FrequencyBasedAnalyzer {
...
+      .select(columnsToGroupBy: _*)
+      .where(atLeastOneNonNullGroupingColumn)
+      .transform(filterOptional(where))
+      .count()
 
     FrequenciesAndNumRows(frequencies, numRows)
   }
+
+  private def filterOptional(where: Option[String])(df: DataFrame) : DataFrame = {
+    if (where.isDefined) df.filter(where.get) else df
```
[the remark this answers is not in the document]
Comment (rambrus) [reply src/main/scala/com/amazon/deequ/analyzers/GroupingAnalyzers.scala]: fixed

File: src/test/scala/com/amazon/deequ/analyzers/runners/AnalysisRunnerTests.scala [hunk 2]
```diff
@@ -185,7 +185,7 @@ class AnalysisRunnerTests extends WordSpec with Matchers with SparkContextSpec w
 
         assert(exception.getMessage == "Could not find all necessary results in the " +
           "MetricsRepository, the calculation of the metrics for these analyzers " +
-          "would be needed: Uniqueness(List(item, att2)), Size(None)")
+          "would be needed: Uniqueness(List(item, att2),None), Size(None)")
```
[the remark this answers is not in the document]
Comment (rambrus) [reply src/test/scala/com/amazon/deequ/analyzers/runners/AnalysisRunnerTests.scala]: Actually, this is the format we've got in the exception message. Had a quick look and found that scala's `toString` method does not put spaces between case class params.
For example:
```
case class DummyClass(a: Int, b: Int, c: Seq[Int])

DummyClass(1, 2, Seq(1, 2)).toString
```
produces:
`res0: String = DummyClass(1,2,List(1, 2))`

**Comment by rambrus** [comment]
@sscdotopen added the missing test cases, please take a look.

**Comment by sscdotopen** [comment]
Thank you for this awesome contribution! Much appreciated!

**Pull request merged**